src/coreclr/jit/morph.cpp

```diff
@@ -14738,6 +14714,40 @@ GenTree* Compiler::fgMorphModToSubMulDiv(GenTreeOp* tree)
     return sub;
 }
 
+//------------------------------------------------------------------------
+// fgMorphUModToAndSub: Transform a % b into the equivalent a & (b - 1).
+// '%' must be unsigned (GT_UMOD).
+// 'a' and 'b' must be integers.
+// 'b' must be a constant and a power of two.
+//
+// Arguments:
+//    tree - The GT_UMOD tree to morph
+//
+// Returns:
+//    The morphed tree
+//
+// Notes:
+//    This is more optimized than calling fgMorphModToSubMulDiv.
+//
+GenTree* Compiler::fgMorphUModToAndSub(GenTreeOp* tree)
+{
+    JITDUMP("\nMorphing UMOD [%06u] to And/Sub\n", dspTreeID(tree));
+
+    assert(tree->OperIs(GT_UMOD));
+    assert(tree->gtOp2->IsIntegralConstAbsPow2());
+
+    var_types type = tree->TypeGet();
+
+    GenTree* const sub    = gtNewOperNode(GT_SUB, type, tree->gtOp2, gtNewOneConNode(type));
```

Comment:
Is there any idea why we can't create a const tree with value `b - 1` here instead?

Reply:
i.e. the resulting tree would be `and(op1, cns(op2.value - 1))`

Reply:
What the resulting codegen look for the tree? Is it the expected `and dstReg, op1Reg, #(op2.Value - 1)`?

Reply:
> Is there any idea why we can't create a const tree with value b - 1 here instead?

I was just thinking that recently. I actually did this very thing for part 2 of 'a % b' optimizations.